infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java

```diff
@@ -290,6 +278,31 @@ private void exposeNonSecureServers(
     }
   }
 
+  private Service createService(
+      Map<String, ServerConfig> internalServers,
+      Map<String, ServerConfig> externalServers,
+      Map<String, ServicePort> unsecuredPorts,
+      Map<String, ServicePort> securedPorts) {
+    Map<String, ServerConfig> allInternalServers = new HashMap<>(internalServers);
+    // Adding external secure services, as they, in fact, internal also.
+    for (ServicePort servicePort : securedPorts.values()) {
+      Map<String, ServerConfig> securedExternalServers = match(externalServers, servicePort);
+      allInternalServers.putAll(securedExternalServers);
+    }
+    Service service =
+        new ServerServiceBuilder()
+            .withName(generate(SERVER_PREFIX, SERVER_UNIQUE_PART_SIZE) + '-' + machineName)
+            .withMachineName(machineName)
+            .withSelectorEntry(CHE_ORIGINAL_NAME_LABEL, pod.getMetadata().getName())
+            .withPorts(new ArrayList<>(unsecuredPorts.values()))
+            .withServers(allInternalServers)
+            .build();
+
+    String serviceName = service.getMetadata().getName();
+    k8sEnv.getServices().put(serviceName, service);
```

Comment:
can we please move this to add to infra to `expose` method so it's not so hidden here in private method?

Reply:
\+